File: app/src/main/java/dev/thehale/papermc_plugin_template/commands/BaseRootCommandModule.java
package dev.thehale.papermc_plugin_template.commands;

import java.util.List;

import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

public class BaseRootCommandModule implements SheepMergeCommandModule {

    @FunctionalInterface
    public interface RootCommandExecutor {
        boolean execute(Player player, String[] args);
    }

    @FunctionalInterface
    public interface RootTabCompleter {
        List<String> complete(CommandSender sender, String[] args);
    }

    private final String root;
    private final RootCommandExecutor executor;
    private final RootTabCompleter tabCompleter;

    protected BaseRootCommandModule(String root, RootCommandExecutor executor, RootTabCompleter tabCompleter) {
        this.root = root;
        this.executor = executor;
        this.tabCompleter = tabCompleter;
    }

    @Override
    public String root() {
        return root;
    }

    @Override
    public boolean execute(Player player, String[] args) {
        return executor.execute(player, args);
    }

    @Override
    public List<String> tabComplete(CommandSender sender, String[] args) {
        return tabCompleter.complete(sender, args);
    }
}

File: app/src/main/kotlin/dev/x208/sheepmerge/commands/BaseRootCommandModule.kt
package dev.x208.sheepmerge.commands

import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

open class BaseRootCommandModule(
    private val root: String,
    private val executor: RootCommandExecutor,
    private val tabCompleter: RootTabCompleter
) : SheepMergeCommandModule {

    fun interface RootCommandExecutor {
        fun execute(player: Player, args: Array<String>): Boolean
    }

    fun interface RootTabCompleter {
        fun complete(sender: CommandSender, args: Array<String>): List<String>
    }

    override fun root(): String {
        return root
    }

    override fun execute(player: Player, args: Array<String>): Boolean {
        return executor.execute(player, args)
    }

    override fun tabComplete(sender: CommandSender, args: Array<String>): List<String> {
        return tabCompleter.complete(sender, args)
    }
}

typealias RootCommandExecutor = BaseRootCommandModule.RootCommandExecutor
typealias RootTabCompleter = BaseRootCommandModule.RootTabCompleter
